### PR TITLE
[RFC] [Relay] Automatic Mixed Precision Pass

### PR DESCRIPTION
**Pull request by AndrewZhaoLuo**
Relevant Links: 

https://discuss.tvm.apache.org/t/rfc-relay-fp32-fp16-model-support/9994
  - Old discussion before the new RFC process was rolled out

https://github.com/apache/tvm/pull/8069
  - Initial PR
  
cc @hogepodge @mbrookhart @anijain2305 @masahi 

Link to tracking issue: https://github.com/apache/tvm/issues/8296

### CODE REVIEW

**Comment by comaniac** [comment]
Thanks for the RFC. I have two questions:
1. How to mark/set the color (i.e., attribute) of every operator?
2. It seems to me that if we register a casting checker instead of just a label (color), then we can simplify the algorithm a lot. Taking the case `A(green) - B(gray) - C(green)` as an example, if we could register a casting rule of B as follows, then we just need one traverse to know if we need cast around B:

    ```
    def amp_B(expr, args):
        a = args[0]
        if (a.dtype is float16):
          return fp16
        return fp32
    ```

    After all, we only need the previous nodes to determine 1) whether to use FP16 implementation, and 2) whether to insert casts. It seems to me that this pass is similar to the layout conversion pass, which uses one traverse to finish everything, so it might be possible for AMP too.

**Comment by AndrewZhaoLuo** [comment]
> Thanks for the RFC. I have two questions:
> 
> 1. How to mark/set the color (i.e., attribute) of every operator?
> 2. It seems to me that if we register a casting checker instead of just a label (color), then we can simplify the algorithm a lot. Taking the case `A(green) - B(gray) - C(green)` as an example, if we could register a casting rule of B as follows, then we just need one traverse to know if we need cast around B:
>    ```
>    def amp_B(expr, args):
>        a = args[0]
>        if (a.dtype is float16):
>          return fp16
>        return fp32
>    ```
>    
>    
>        
>          
>        
>    
>          
>        
>    
>        
>      
>    After all, we only need the previous nodes to determine 1) whether to use FP16 implementation, and 2) whether to insert casts. It seems to me that this pass is similar to the layout conversion pass, which uses one traverse to finish everything, so it might be possible for AMP too.

Yep that is correct. This RFC has an initial PR here: https://github.com/apache/tvm/pull/8069.

To answer your questions:
1.  src/relay/transforms/fp32_to_fp16.h -- DefaultFP16Colorer is the default way. But the only thing we need is a callable with type CallNode*(Color). So you could write your own colorer that does arbitrary stuff when only looking at a single node at a time.

2. This is functionally what is done in the PR I link.

**Comment by AndrewZhaoLuo** [comment]
> Thanks for the answers. I'll review the PR to get more implementation details.
> One more question regarding the extensibility: can this be extended easily to support bfloat16?

It should be trivial (hope I don't eat my words). I'm not 100% sure of the support for bfloat16 in current relay ops however.

**Comment by AndrewZhaoLuo** [comment]
I don't know Chris Sullivan's github handle so if someone could cc him too that would be great.

**Comment by tmoreau89** [comment]
CCing @csullivan

**Comment by comaniac** [comment]
> > Thanks for the answers. I'll review the PR to get more implementation details.
> > One more question regarding the extensibility: can this be extended easily to support bfloat16?
> 
> It should be trivial (hope I don't eat my words). I'm not 100% sure of the support for bfloat16 in current relay ops however.

TVM has limited bfloat16 support now but it's on the way, so it would be better for this RFC to also consider this case, even the initial version may not cover it.

**Comment by AndrewZhaoLuo** [comment]
So the associated PR is getting closer to a mergeable state. Is this RFC ready to merge?

**Comment by tqchen** [comment]
cc @comaniac would be great if you can help shepherd this RFC

File: rfcs/0001-AMP_pass.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+- Why is this design the best in the space of possible designs?
+
+Other alternatives require a lot more work and changes and could probably considered future goals of TVM.
+This include automatic mixed precision training.
+
+- What other designs have been considered and what is the rationale for not choosing them?
+
+We can support automatic mixed precision retraining though that is a much, much larger future goal. It's
+good to have this in the meantime.
```
Comment (comaniac) [inline rfcs/0001-AMP_pass.md:103]: The answer to this question should come with a discussion of existing mechanisms used by other frameworks, such as XLA and PyTorch.

File: rfcs/0001-AMP_pass.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+- What is the impact of not doing this?
+
+TVM is not the best tool for making models go fast as we leave a lot of free speedup on the table.
+
+# Prior art
+[prior-art]: #prior-art
```
Comment (comaniac) [inline rfcs/0001-AMP_pass.md:110]: It's better to also discuss the tensor cache mechanism used in PyTorch.

File: rfcs/0001-AMP_pass.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+Many machine learning models can move significant portions of their computational graphs into the FP16 space 
+without significant loss of accuracy. For many pieces of hardware this also comes with a boost in speed. In 
+the past utilizing FP16 in mixed precision training saw significant [increases in convergence speed](https://pytorch.org/blog/accelerating-training-on-nvidia-gpus-with-pytorch-automatic-mixed-precision/). 
+
+We should expect similar increases for inference. This speed increase without accuracy loss is highly desirable
+for many users.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
```
Comment (comaniac) [inline rfcs/0001-AMP_pass.md:33]: It would be better to provide an example at the end of this section. i.e., how this pass is used and what's the result IR looks like.

File: rfcs/0001-AMP_pass.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+a function, which maps operation types like `conv2d` to an accumulation datatype as well as an output 
+datatype. The output datatype is the type other operations down the line will likely ingest from the previous
+calculation while the accumulation datatype describes the size of buffer where the results are initially
+stored. For NVidia's tensor cores for example many operations accumulate in FP32 but have an output datatype
+of FP16. The default implementation will follow this guideline closely and will by default have all 
+operations output FP16 and accumulate in FP32 only if TVM supports mixed datatypes for that particular
+operation.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
```
Comment (comaniac) [inline rfcs/0001-AMP_pass.md:66]: This section should also discuss the implementation. Specifically, 1) the interface of annotating an op with color, 2) the coloring algorithm in the pass, 3) some corner cases (i.e., ops) that need more care.

File: rfcs/0001-AMP_pass.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+Many of the ideas are taken from Tensorflow's [automatic mixed precision training framework](https://on-demand.gputechconf.com/gtcdc/2019/pdf/dc91247-automatic-mixed-precision-in-tensorflow.pdf)
+and the initial "Green", "Gray", and "Red" lists are based [similarly](github.com/tensorflow/tensorflow/blob/v2.5.0/tensorflow/core/grappler/optimizers/auto_mixed_precision_lists.h). 
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+
+We still need to make sure that the current design and knobs exposed provide extensibility to every hardware platform out there.
```
Comment (comaniac) [inline rfcs/0001-AMP_pass.md:120]: This seems not the answer to this question. IIUC, the initial implementation of the pass has been merged, so we should mention that here with the PR link.

File: rfcs/0001-AMP_pass.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+
+Probably a lot of edge cases of operations within TVM.
```
Comment (comaniac) [inline rfcs/0001-AMP_pass.md:124]: This is too vague. It's better to have a quantitative metric toward to a stable release. For example, you can set up a benchmark with a set of models, and the goal is to make all of them work well with AMP (in terms of the performance and accuracy) on both CPU and GPU.

In addition, it would be better to also investigate how AutoScheduler works with AMP models. Since tuning is an important feature in TVM, the impact of AMP would be moderated if a tuned FP32 model can still run faster than an AMP model.

File: rfcs/0001-AMP_pass.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
+- Feature Name: Automatic Mixed Precision Pass
+- Start Date: 2021-06-08 
+- RFC PR: TODO
```
Comment (comaniac) [inline rfcs/0001-AMP_pass.md:3]: Update the RFC PR.

File: rfcs/0001-AMP_pass.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+- GitHub Issue: TODO
```
Comment (comaniac) [inline rfcs/0001-AMP_pass.md:4]: As this RFC is guaranteed to be merged and the feature must be landed, It should be fine to open a tracking issue now and update the link here.

**Review by comaniac — CHANGES_REQUESTED** [review]
The concept and algorithm look good to me, but it would be better to provide more implementation/design details.

File: rfcs/0001-AMP_pass.md (hunk 8 above)
Comment (comaniac) [reply rfcs/0001-AMP_pass.md:4]: I found that the issue has been opened, so please just update the link here.

**Comment by AndrewZhaoLuo** [comment]
Thanks for driving this review @comaniac. I'll get to this later in the week.

File: rfcs/0001-AMP_pass.md (hunk 4 above)
Comment (u99127) [reply rfcs/0001-AMP_pass.md:66]: The M6G for instance also has FP16 instructions from the AArch64 ISA including Advanced SIMD fixed length 128 bit vectors- perhaps also investigate the suitability of the representation to support more than 1 backend with this ? I also expect that with the SVE (Scalable Vector extensions) you'd see FP16 instructions there as well but that's a different kettle of fish.

I also expect that in the uTVM case it would be interesting in a later date for the MVE instruction set with FP16 being present there as well. I'd like to look at whether there is an multiply and accumulate instruction there that matches with an FP32 accumulator.  Off hand I'm not sure about the answer to that question.

Just something to consider.

My 2 cents 

Ramana

Comment (comaniac) [reply rfcs/0001-AMP_pass.md:66]: Thanks for the suggestion. I agree that it would be interesting to see how AMP works on m6g instances. Maybe we can document it in future possibilities section so that everyone interested in this feature can potentially take over after the AMP pass is stable on x86 platforms.

**Comment by AndrewZhaoLuo** [comment]
Going to get to this tomorrow 😬. Promise 🤞

**Comment by comaniac** [comment]
btw, according to #17, please update the RFC number on the file name to align with this PR number.

**Comment by comaniac** [comment]
Took a quick pass to the updated RFC. I think it's almost ready to merge as long as the last 3 comments are resolved.

File: rfcs/0001-AMP_pass.md (hunk 1 above)
Comment (AndrewZhaoLuo) [reply rfcs/0001-AMP_pass.md]: Done. Please let me know if this is sufficient. Don't have the best background on some of this stuff.

File: rfcs/0001-AMP_pass.md (hunk 2 above)
Comment (AndrewZhaoLuo) [reply rfcs/0001-AMP_pass.md]: I am not familiar with this and cannot find a good article, do you have a link?

File: rfcs/0001-AMP_pass.md (hunk 4 above)
Comment (AndrewZhaoLuo) [reply rfcs/0001-AMP_pass.md]: Hey all, in the case where does not have an instruction for accumulating fp16 results in fp32, the user has the interface to turn off accumulating in fp32. 

Right now the default is designed with NVidia GPU in mind, but example pass settings can be easily made.

As for concerns about the representation being suitable for your device, do you have any specific concerns? As long as the codegen path to the device is good and clear I do not think the representation will be a problem.

Comment (AndrewZhaoLuo) [reply rfcs/0001-AMP_pass.md]: @jwfromm I believe we also had good results running mixed precision on M6g (graviton) instances?

Comment (AndrewZhaoLuo) [reply rfcs/0001-AMP_pass.md]: @comaniac x86 is not really the target for this pass as it has poor hardware FP16 support. I'll add a list of supported targets we will aim for.

**Comment by AndrewZhaoLuo** [comment]
PTAL @comaniac

File: rfcs/0001-AMP_pass.md (hunk 4 above)
Comment (comaniac) [reply rfcs/0001-AMP_pass.md]: Yeah that would be sufficient for now. We will also need the BF16 support as well.

File: rfcs/0001-AMP_pass.md (hunk 2 above)
Comment (comaniac) [reply rfcs/0001-AMP_pass.md]: It's by tracing the PyTorch code so I don't have an article talking about this. The basic idea is since PyTorch uses interpreter to execute the model, it dynamically casts a FP32 tensor to FP16 when the op it's going to execute requires FP16 inputs. Accordingly, to avoid casting a tensor from FP32 to FP16 multiple times, it has a tensor cache to cache the casted tensors, so that they can be directly used in the future.

With the above content it should be clear that why TVM should adopt the XLA-like approach instead of PyTorch, because TVM is more compile-oriented and we are able to embrace more graph-level optimizations.

Comment (AndrewZhaoLuo) [reply rfcs/0001-AMP_pass.md]: Done, added discussion on this topic

**Comment by AndrewZhaoLuo** [comment]
@comaniac, I'll be talking about this at the TVM community meeting tomorrow so put off merging until after.

**Review by comaniac — APPROVED** [review]
LGTM. Will merge after the community meeting if there's no objection.

**Comment by AndrewZhaoLuo** [comment]
If there is not other objections, this will be merged on monday.

**Comment by comaniac** [comment]
Thanks @AndrewZhaoLuo

**Comment by MeJerry215** [comment]
@AndrewZhaoLuo will it remove cast weight to float16 from graph? and make weight as float16 when build lib. 
in my opinion, it will reduce the bandwidth.

**Comment by masahi** [comment]
@MeJerry215 Yes, casting of weight to fp16 is done at compile time by `FoldConstant` pass, so weights will be in fp16 at deploy time.

**Comment by gayatripk1** [comment]
@AndrewZhaoLuo, When we enable mixed precision arithmetic, the output types of a model depends on the last operation in the relay IR, ie, if it convolution, it will be in float16 and for the case argmax it will be int32 and in the case of softmax it will be float32. So, is this expected? Or should the nodes that are marked as output nodes should be treated differently so that they return the expected specific out_type irrespective of the optimization(mixed precision pass) happening? Please let me know your thoughts on this.
